### DISCUSSION TITLE
Mux and sysex

### DISCUSSION THREAD

**Discussion by klausnielsen** [tttapa/Control-Surface#1011] [General]
Hi everyone,
I'm working on a midi controller that will send out sysex to an old 80's synth. Using [an example](https://forum.arduino.cc/t/midi-shield-sending-sysex/690648/2) Pieter posted on Arduino Forum I can very easily get a single 10K potentiometer (hooked up to +5v, A0, GND) to send sysex in realtime (see below).
But since I will be needing a LOT of pots, I'm thinking I need to use multiplexers similar to the example on CCPotentiometers given in section 4 of the ['Getting Started' documentation](https://tttapa.github.io/Control-Surface-doc/Doxygen/d5/d7d/md_pages_Getting-Started.html). I've searched high an low for examples on how to do this, but without luck. How would you go about muxing this sysex sketch?
Hope someone has some ideas
Cheers
```C
// Include the library
#include <Control_Surface.h>

// Select the serial port to use.
auto &serial = Serial;

// Instantiate a Serial MIDI interface at the default MIDI baud rate.
SerialMIDI_Interface<decltype(serial)> midi {serial, MIDI_BAUD};

// Reads, filters and detects changes of an analog input
// Produces a 7-bit output [0-127]
FilteredAnalog<7> potentiometer { A0 }; 

// MIDI System Exclusive data to send - VCF Freq
uint8_t sysex_data[] {
  0xF0, 0x10, 0x06, 0x06, 0x15, 0x00, 0xF7,
};
// Reference to easily update the parameter value of the SysEx message
uint8_t &vcf_value = sysex_data[5];

void setup() {
  // Initialize MIDI interface and analog input
  midi.begin();
  potentiometer.setupADC();
}

void loop() {
  midi.update(); // Handle and discard MIDI input

  // Check whether the potentiometer value changed
  if (potentiometer.update()) {
    vcf_value = potentiometer.getValue(); // if it did, update the message
    SysExMessage msg { sysex_data, sizeof(sysex_data) };
    midi.send(msg); // and send it out over MIDI
  }
}
```

**Comment by tttapa** [comment]
You could create an array of `FilteredAnalog`, and use `mux.pin(p)` instead of `A0`. Then simply loop over all potentiometers and send the appropriate value using a for loop.

```cpp
CD74HC4067 mux { ... };

FilteredAnalog<7> potentiometers[] {
  mux.pin(0), mux.pin(1), // ...
};
```

Some examples that could be useful:
- https://tttapa.github.io/Control-Surface-doc/Doxygen/d2/d96/3_8FilteredAnalogReadSerial_8ino-example.html
- https://tttapa.github.io/Control-Surface-doc/Doxygen/d4/d3b/Custom-MIDI-Output-Element_8ino-example.html

**Reply by klausnielsen** [reply]
Thanks, Pieter, for taking the time to answer all of these questions. I see your point about creating an array of `FilteredAnalog` with the mux pins, but I'm having a hard time figuring out how to assign a different sysex address to each mux pin (which was super easy and smooth with `CCPotentiometer`.)
I'm sure the answers is somewhere in the two examples you reference, but I cannot crack the nut. I would be super grateful if you could please show me how you would make for instance `mux.pin(0)` send out `sysex_data[5]` values for the sysex string {0xF0, 0x10, 0x06, 0x06, **_0x15_**, 0x00, 0xF7} and `mux.pin(1)` send out `sysex_data[5]` values for {0xF0, 0x10, 0x06, 0x06, **_0x16_**, 0x00, 0xF7}. (Index[4] of course being the sysex parameter ID.) I could hopefully pick up from there.

**Reply by tttapa** [reply]
If you want something similar to `CCPotentiometer`, you could try something like this (untested):

```cpp
#include <Control_Surface.h>

uint8_t sysex_data[] {
  0xF0, 0x10, 0x06, 0x06, 0x15, 0x00, 0xF7,
};

class SysExPot : public MIDIOutputElement {
 public:
  SysExPot(pin_t pin, uint8_t param_id)
    : pot(pin), param_id(param_id) {}

  void begin() final override {
    pot.resetToCurrentValue();
  }

  void update() final override {
    if (pot.update())
      send(pot.getValue());
  }

  void send(uint8_t value) {
    sysex_data[4] = param_id;
    sysex_data[5] = value;
    Control_Surface.send(SysExMessage{sysex_data});
  }

 private:
  FilteredAnalog<7> pot;
  uint8_t param_id;
};

auto &serial = Serial;
SerialMIDI_Interface<decltype(serial)> midi {serial, MIDI_BAUD};

CD74HC4067 mux {A0, {3, 4, 5, 6}};

SysExPot pots[] {
  {mux.pin(0), 0x15},
  {mux.pin(1), 0x16},
};

void setup() {
  Control_Surface.begin();
}

void loop() {
  Control_Surface.loop();
}
```
See the Custom-MIDI-Output-Element example from my previous post for more information about the functions I defined in the `SysExPot` class.

**Reply by klausnielsen** [reply]
Fantastic! It works right out of the box :)
Thanks Pieter - I really appreciate your help.
I will be sure to test this in the coming weeks - pushing it as far I can. (Will be needing 64 pots or so + a number of push buttons.)
Already I see the next challenge: Some pots need to go from 0-127, while others will need to go from 0-63 or from -63 to +63(!) I will have some stuff to learn :)
